{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9197b49b-54ae-4129-8864-cb68154b2bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "from pathlib import Path\n",
    "import json\n",
    "\n",
    "#### functions for ultraplex statistics extraction ####\n",
    "def extract_nread_with_barcode(s):\n",
    "    ''' extract nread with barcode from ultraplex.log'''\n",
    "    pattern = r\"(\\d+)\\s+\\((\\d+\\.\\d+)%\\)\\s+reads\\s+correctly\\s+assigned\"\n",
    "    match = re.search(pattern, s)\n",
    "    if match:\n",
    "        num_reads = match.group(1)\n",
    "        percentage = match.group(2)\n",
    "        return num_reads, percentage\n",
    "    else:\n",
    "        print(s)\n",
    "        return None, None\n",
    "def extract_input_reads(s):\n",
    "    '''Demultiplexing complete! 21532666 reads processed in 1095.0 seconds'''\n",
    "    pattern = r'(?<=Demultiplexing complete! )\\d+(?= reads processed)'\n",
    "    \n",
    "    matches = re.findall(pattern, s)\n",
    "\n",
    "    if matches:\n",
    "        num_reads = int(matches[0])\n",
    "        \n",
    "        return num_reads\n",
    "    else:\n",
    "        print(s)\n",
    "        return None\n",
    "\n",
    "def extract_quality_trimmed(s):\n",
    "    pattern = r\"(\\d+)\\s+\\((\\d+\\.\\d+)%\\)\\s+reads\\s+quality\\s+trimmed\"\n",
    "    match = re.search(pattern, s)\n",
    "    if match:\n",
    "        num_reads = match.group(1)\n",
    "        percentage = match.group(2)\n",
    "        return num_reads, percentage\n",
    "    else:\n",
    "        print(s)\n",
    "        return None, None\n",
    "\n",
    "def get_ultraplex_stat(basedir):\n",
    "    ultraplex_stat = []\n",
    "    for file in basedir.glob('*/fastqs/ultraplex*.log'):\n",
    "        libname = file.parent.parent.name\n",
    "        with open(file) as f:\n",
    "            lines = f.readlines()\n",
    "            barcode_nread, barcode_perc = extract_nread_with_barcode(lines[-1])\n",
    "            ultraplex_input_nread = extract_input_reads([l for l in lines if 'Demultiplex' in l][0])\n",
    "        ultraplex_stat.append([libname, barcode_nread, barcode_perc, ultraplex_input_nread])\n",
    "    ultraplex_stat = pd.DataFrame(ultraplex_stat, columns = ['libname', 'nread_w_barcode', '%barcode', 'input_to_ultraplex']\n",
    "                        ).drop_duplicates('libname').set_index('libname')\n",
    "    return ultraplex_stat\n",
    "\n",
    "def get_fastp_stat(basedir):\n",
    "    ''' extracts fastp statistics '''\n",
    "    fastp_stats = []\n",
    "    for fastp_stat_file in basedir.glob('QC/*umi.json'):\n",
    "        fastp_stat = json.load(open(fastp_stat_file))\n",
    "        fastp_stats.append([fastp_stat_file.name.split('.')[0],\n",
    "                        fastp_stat['summary']['before_filtering']['total_reads'],\n",
    "                        fastp_stat['summary']['after_filtering']['total_reads']]\n",
    "                        )\n",
    "    fastp_stats = pd.DataFrame(fastp_stats, columns = ['libname', 'before_umi_polyG', 'after_umi_polyG']).set_index('libname')\n",
    "    fastp_stats['%polyG or too short']=100*(fastp_stats['before_umi_polyG']-fastp_stats['after_umi_polyG'])/fastp_stats['before_umi_polyG']\n",
    "    return fastp_stats\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d60b2b5-fd7b-4fa7-9ebf-d8ced0b9f299",
   "metadata": {},
   "outputs": [],
   "source": [
    "basedir = Path('/tscc/projects/ps-yeolab3/bay001/for_charlene/scratch/sra_test')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0859e980-b1db-4018-8b5c-22008deaf25d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tscc/nfs/home/hsher/miniconda3/envs/my_metadensity/lib/python3.7/site-packages/ipykernel_launcher.py:22: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.\n"
     ]
    }
   ],
   "source": [
    "# read inputs:\n",
    "cutadapt = pd.read_csv(basedir/'QC/cutadapt_stat.csv', index_col = 0)\n",
    "genome_stat = pd.read_csv(basedir/'QC/mapping_stats.csv', index_col = 0)\n",
    "dup_df = pd.read_csv(basedir/'QC/dup_level.csv', index_col = 0)\n",
    "ultraplex_stat = get_ultraplex_stat(Path('.'))\n",
    "fastp_stats = get_fastp_stat(Path('.'))\n",
    "\n",
    "# generate mapping summary\n",
    "genome_stat['libname']=genome_stat['STAR Log filename'].str.split('/', expand = True)[0]\n",
    "mapping_summary = pd.concat([genome_stat.groupby(by = 'libname')['Uniquely mapped reads number'].sum(),\n",
    "        genome_stat.groupby(by = 'libname')['Number of reads mapped to multiple loci'].sum(),\n",
    "                            genome_stat.groupby(by = 'libname')['Number of input reads'].sum(),\n",
    "        ], axis = 1\n",
    "        )\n",
    "mapping_summary['total_mapped']=mapping_summary[['Uniquely mapped reads number', 'Number of reads mapped to multiple loci']].sum(axis = 1)\n",
    "mapping_summary['%Unique map']=100*mapping_summary['Uniquely mapped reads number']/mapping_summary['Number of input reads']\n",
    "mapping_summary['%Multi map']=100*mapping_summary['Number of reads mapped to multiple loci']/mapping_summary['Number of input reads']\n",
    "mapping_summary['% mapped']=100*mapping_summary['total_mapped']/mapping_summary['Number of input reads']\n",
    "\n",
    "# generate dedup summary: is the sum of read 1 and read2\n",
    "dup_df['libname']=dup_df['dup_bam'].str.split('/', expand = True)[0]\n",
    "dup_summary_df=dup_df.groupby(by = ['libname'])['before_dedup', 'after_dedup'].sum()\n",
    "dup_summary_df['% Unique frag']=100*dup_summary_df['after_dedup']/dup_summary_df['before_dedup']\n",
    "\n",
    "# generate repeat counts summary\n",
    "repeat_cnt_stat = []\n",
    "for repeat_cnt_file in list((basedir/'counts/repeats/megatables/class').glob('*tsv.gz')):\n",
    "    repeat_cnt = pd.read_csv(repeat_cnt_file, sep = '\\t', index_col = 0)\n",
    "    n_read_in_repeat = repeat_cnt.sum().sum()\n",
    "    n_read_in_rRNA = repeat_cnt.loc['rRNA'].sum()\n",
    "    \n",
    "    repeat_cnt_stat.append([repeat_cnt_file.name.split('.')[0],\n",
    "                            n_read_in_repeat,\n",
    "                            n_read_in_rRNA]\n",
    "                        )\n",
    "repeat_cnt_stat = pd.DataFrame(repeat_cnt_stat, columns = ['libname', 'n_read_in_repeat', 'n_read_in_rRNA']).set_index('libname')\n",
    "\n",
    "# generate genome counts summary\n",
    "reads_in_window = []\n",
    "for region_cnt in list(Path(basedir/'QC/read_count').glob('*region.csv')):\n",
    "    region_read_count = pd.read_csv(region_cnt, index_col = 0)\n",
    "    reads_in_window.append([region_cnt.name.split('.')[0], region_read_count.sum().sum()])\n",
    "reads_in_window = pd.DataFrame(reads_in_window, columns = ['libname', 'nread_in_genome_window']).set_index('libname')\n",
    "\n",
    "# make summary\n",
    "cutadapt.index = [i.split('.')[0] for i in cutadapt.index]\n",
    "try:\n",
    "    summary = pd.concat([cutadapt[['Total reads processed', '% Reads that were too short']],\n",
    "            fastp_stats,\n",
    "            ultraplex_stat,\n",
    "\n",
    "            mapping_summary,\n",
    "            dup_summary_df,\n",
    "\n",
    "            reads_in_window,\n",
    "                        repeat_cnt_stat],\n",
    "            axis = 1)\n",
    "    is_paired = False\n",
    "except:\n",
    "    summary = pd.concat([cutadapt[['Total read pairs processed', '% Pairs that were too short']],\n",
    "            fastp_stats,\n",
    "            ultraplex_stat,\n",
    "\n",
    "            mapping_summary,\n",
    "            dup_summary_df,\n",
    "\n",
    "            reads_in_window,\n",
    "                        repeat_cnt_stat],\n",
    "            axis = 1)\n",
    "    is_paired = True\n",
    "\n",
    "if not is_paired:\n",
    "    summary['%_in_genome_window']=100*summary['nread_in_genome_window']/summary['after_dedup']\n",
    "    summary['%_in_repeat']=100*summary['n_read_in_repeat']/summary['after_dedup']\n",
    "    summary['%_in_rRNA']=100*summary['n_read_in_rRNA']/summary['after_dedup']\n",
    "else:\n",
    "    # counting is only counting 1 read, but samtools counts both\n",
    "    summary['%_in_genome_window']=100*2*summary['nread_in_genome_window']/summary['after_dedup']\n",
    "    summary['%_in_repeat']=100*2*summary['n_read_in_repeat']/summary['after_dedup']\n",
    "    summary['%_in_rRNA']=100*2*summary['n_read_in_rRNA']/summary['after_dedup']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "58d89fa7-4066-4efa-ae92-cf6ccc6e5052",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Total reads processed</th>\n",
       "      <th>% Reads that were too short</th>\n",
       "      <th>before_umi_polyG</th>\n",
       "      <th>after_umi_polyG</th>\n",
       "      <th>%polyG or too short</th>\n",
       "      <th>nread_w_barcode</th>\n",
       "      <th>%barcode</th>\n",
       "      <th>input_to_ultraplex</th>\n",
       "      <th>Uniquely mapped reads number</th>\n",
       "      <th>Number of reads mapped to multiple loci</th>\n",
       "      <th>...</th>\n",
       "      <th>% mapped</th>\n",
       "      <th>before_dedup</th>\n",
       "      <th>after_dedup</th>\n",
       "      <th>% Unique frag</th>\n",
       "      <th>nread_in_genome_window</th>\n",
       "      <th>n_read_in_repeat</th>\n",
       "      <th>n_read_in_rRNA</th>\n",
       "      <th>%_in_genome_window</th>\n",
       "      <th>%_in_repeat</th>\n",
       "      <th>%_in_rRNA</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>K562_rep4</th>\n",
       "      <td>236070272</td>\n",
       "      <td>10.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36119518.0</td>\n",
       "      <td>144887779.0</td>\n",
       "      <td>...</td>\n",
       "      <td>90.642654</td>\n",
       "      <td>181007297</td>\n",
       "      <td>11655796</td>\n",
       "      <td>6.439407</td>\n",
       "      <td>7740556</td>\n",
       "      <td>3380094</td>\n",
       "      <td>574876</td>\n",
       "      <td>66.409501</td>\n",
       "      <td>28.999255</td>\n",
       "      <td>4.932104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K562_rep6</th>\n",
       "      <td>316340455</td>\n",
       "      <td>3.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>45723494.0</td>\n",
       "      <td>220340909.0</td>\n",
       "      <td>...</td>\n",
       "      <td>91.142603</td>\n",
       "      <td>266064403</td>\n",
       "      <td>8281139</td>\n",
       "      <td>3.112457</td>\n",
       "      <td>4996487</td>\n",
       "      <td>2800053</td>\n",
       "      <td>573098</td>\n",
       "      <td>60.335746</td>\n",
       "      <td>33.812414</td>\n",
       "      <td>6.920521</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Total reads processed  % Reads that were too short  \\\n",
       "K562_rep4              236070272                         10.4   \n",
       "K562_rep6              316340455                          3.5   \n",
       "\n",
       "          before_umi_polyG after_umi_polyG %polyG or too short  \\\n",
       "K562_rep4              NaN             NaN                 NaN   \n",
       "K562_rep6              NaN             NaN                 NaN   \n",
       "\n",
       "          nread_w_barcode %barcode input_to_ultraplex  \\\n",
       "K562_rep4             NaN      NaN                NaN   \n",
       "K562_rep6             NaN      NaN                NaN   \n",
       "\n",
       "           Uniquely mapped reads number  \\\n",
       "K562_rep4                    36119518.0   \n",
       "K562_rep6                    45723494.0   \n",
       "\n",
       "           Number of reads mapped to multiple loci  ...   % mapped  \\\n",
       "K562_rep4                              144887779.0  ...  90.642654   \n",
       "K562_rep6                              220340909.0  ...  91.142603   \n",
       "\n",
       "           before_dedup  after_dedup  % Unique frag  nread_in_genome_window  \\\n",
       "K562_rep4     181007297     11655796       6.439407                 7740556   \n",
       "K562_rep6     266064403      8281139       3.112457                 4996487   \n",
       "\n",
       "           n_read_in_repeat  n_read_in_rRNA  %_in_genome_window  %_in_repeat  \\\n",
       "K562_rep4           3380094          574876           66.409501    28.999255   \n",
       "K562_rep6           2800053          573098           60.335746    33.812414   \n",
       "\n",
       "           %_in_rRNA  \n",
       "K562_rep4   4.932104  \n",
       "K562_rep6   6.920521  \n",
       "\n",
       "[2 rows x 24 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e681a433-3d04-43d2-9e63-5fd9ae349fb5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "my_metadensity",
   "language": "python",
   "name": "my_metadensity"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
